app/boot_control/common.py

```diff
@@ -0,0 +1,400 @@
+            subprocess_call(_cmd, raise_exception=True)
+        except CalledProcessError as e:
+            if e.returncode == 32 and str(e.stderr).find("not mounted") != -1:
+                if not ignore_unmounted:  # ignore umounted error
+                    raise MountError(f"{target} is not mounted")
+                return
+
+            _failure_msg = f"failed to umount {target}: {e!r}"
+            logger.warning(_failure_msg)
+
+            if not ignore_error:
+                raise MountError(_failure_msg) from e
+
+    @classmethod
+    def mkfs_ext4(cls, dev: str):
+        try:
+            _cmd = f"mkfs.ext4 -F {dev}"
+            subprocess_call(_cmd, raise_exception=True)
+        except CalledProcessError as e:
+            _failure_msg = f"failed to apply mkfs.ext4 on {dev}: {e!r}"
+            logger.error(_failure_msg)
+            raise MkfsError(_failure_msg)
+
+    @classmethod
+    def mount_refroot(
+        cls,
+        standby_slot_dev: str,
+        active_slot_dev: str,
+        refroot_mount_point: str,
+        *,
+        standby_as_ref: bool,
+    ):
+        _refroot_dev = standby_slot_dev if standby_as_ref else active_slot_dev
+
+        # NOTE: set raise_exception to false to allow not mounted
+        # not mounted dev will have empty return str
+        if _refroot_active_mount_point := CMDHelperFuncs.get_mount_point_by_dev(
+            _refroot_dev, raise_exception=False
+        ):
+            _mount_options = ["bind", "ro"]
+            CMDHelperFuncs.mount(
+                _refroot_active_mount_point,
+                refroot_mount_point,
+                _mount_options,
+            )
+        else:
+            # NOTE: refroot is expected to be mounted,
+            # if refroot is active rootfs, it is mounted as /;
+            # if refroot is standby rootfs(in-place update mode),
+            # it will be mounted to /mnt/standby(rw), so we still mount it as bind,ro
+            raise MountError("refroot is expected to be mounted")
+
+
+class GrubABPartitionDetecter:
+    """
+    Expected layout:
+        /dev/sdx
+            - sdx1: dedicated boot partition
+            - sdx2: A partition
+            - sdx3: B partition
+
+    slot_name is the dev name of the A/B partition
+    """
+
+    def __init__(self) -> None:
+        self.active_slot, self.active_dev = self._detect_active_slot()
+        self.standby_slot, self.standby_dev = self._detect_standby_slot(self.active_dev)
+
+    @staticmethod
+    def _get_sibling_dev(slot_dev_path: str) -> str:
+        """
+        Expected partition layout:
+            /dev/sdx
+                - sdx1: dedicated boot partition
+                - sdx2: A partition
+                - sdx3: B partition
+        """
+        parent = CMDHelperFuncs.get_parent_dev(slot_dev_path)
+        family = CMDHelperFuncs.get_dev_family(parent)
+
+        # NOTE: skip the first 2 lines: parent dev and boot dev
+        res = set(family[2:]) - {slot_dev_path}
```

Comment:
`lsblk` always showing partition in the order of actual partition layout, so if we expect partition layout to be:
```
            /dev/sdx
                - sdx1: dedicated boot partition
                - sdx2: A partition
                - sdx3: B partition
```
the first line must be parent dev, and the second line must be boot dev.

Currently if the partition layout is unexpected, currently it will raise a `ABPartitionError`.

Reply:
> the first line must be parent dev, 
That is true, but

> and the second line must be boot dev.
This depends on the /boot partition device number. Current implementation explicitly excludes /boot partition.

Reply:
I think we should explicitly expect partition layout as
```
/dev/sda
    /dev/sda1: boot
    /dev/sda2: slot a
    /dev/sda3: slot b
```
other layout should be treated as an `ABPartitionError`.

Reply:
on x2 main ecu, there 4 partitions:
```
sda
  sda1 vfat /boot/efi
  sda2 ext4 /boot
  sda3 ext4 /
  sda4 ext4 
```
Previous implementation support this structure, so please fix it to support it.

Reply:
AB partition detecting mechanism is updated in [9d87b88](https://github.com/tier4/ota-client/pull/136/commits/9d87b8889cc6fb9528e722cb089fdff875fd7447)
```
* now we assume the last 2 partitions are AB partitions,
  if not the case, an ABPartitionError will be raised.
```